### PR TITLE
Remove deprecated code, update changelog for version 2.3

### PR DESCRIPTION
**Pull request by bhazelton**
## Description

Prepare for version 2.3

Changes include:
- Bump required h5py version to 3.1. This is mostly because we can't actually test against 3.0 because the conda packages don't work, so we've been testing against 3.1 in in our min versions tests. I think it's better to specify the version we actually know works.
- Make some changes in UVCal (related to requiring some new metadata) marked to be changed in version 2.3 and adjust related tests.
- Add version numbers to a few deprecation warnings that were missing specific version numbers for when they would be removed.
- Update the changelog

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation change (documentation changes only)
- [ ] Version change
- [ ] Build or continuous integration change

## Checklist:

- [x] I have read the [contribution guide](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/main/.github/CONTRIBUTING.md).
- [x] My code follows the code style of this project.

Version change checklist:
- [x] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/main/CHANGELOG.md) to put all the unreleased changes under the new version (leaving the unreleased section empty).
- [x] I have noted any dependency changes since the last version and will update the conda package build accordingly.
h5py miniumum bumped to 3.1, lunarsky minimum bumped to 0.2.1

### CODE REVIEW

File: pyuvdata/uvdata/uvdata.py [hunk 1]
```diff
@@ -2225,7 +2225,8 @@ def __setattr__(self, __name, __value) -> None:
             ):
                 warnings.warn(
                     f"phase_center_catalog is already set, so {__name}, which is an "
-                    "old phase attribute, cannot be set.",
+                    "old phase attribute, cannot be set. This will become an error in "
```
Comment (kartographer) [inline pyuvdata/uvdata/uvdata.py:2228]: Maybe worth adding to milestone list if not already.

File: pyuvdata/uvdata/uvdata.py [hunk 2]
```diff
@@ -12565,7 +12566,7 @@ def read(
                 "The `make_multi_phase` option is deprecated. Since all objects now "
                 "support multiple phase centers, the only effect of setting this to "
                 "True is to prevent rephasing. A better approach is to set "
-                "`allow_rephase=False`.",
+                "`allow_rephase=False`. This will become an error in version 3.0.",
```
Comment (kartographer) [inline pyuvdata/uvdata/uvdata.py:12569]: Maybe worth adding to milestone list if not already.

File: pyuvdata/uvdata/uvfits.py [hunk 3]
```diff
@@ -973,7 +973,7 @@ def write_uvfits(
                     "UVFITS-required metadata are now set automatically to their "
                     'best known values, and no longer need to be "spoofed". As such, '
                     "the spoof_nonessential parameter is deprecated, and will be "
-                    "removed in a future release."
+                    "removed in version 2.4."
```
Comment (kartographer) [inline pyuvdata/uvdata/uvfits.py:976]: Maybe worth adding to milestone list if not already.

File: pyuvdata/uvflag/uvflag.py [hunk 4]
```diff
@@ -72,7 +72,7 @@ def lst_from_uv(uv):
     """
     warnings.warn(
         "The lst_from_uv function is deprecated. Use the `set_lsts_from_time_array` "
-        "method on the input object.",
+        "method on the input object. This will become an error in version 2.4.",
```
Comment (kartographer) [inline pyuvdata/uvflag/uvflag.py:75]: Maybe worth adding to milestone list if not already.

**Review by kartographer — APPROVED** [review]
Looks good to me @bhazelton! I had a few GitHub housekeeping comments, but there's nothing in this PR that I see needs changing.

File: pyuvdata/uvdata/uvdata.py (hunk 1 above)
Comment (bhazelton) [reply pyuvdata/uvdata/uvdata.py:2228]: It's there already.

File: pyuvdata/uvdata/uvdata.py (hunk 2 above)
Comment (bhazelton) [reply pyuvdata/uvdata/uvdata.py:12569]: It's there already.

File: pyuvdata/uvdata/uvfits.py (hunk 3 above)
Comment (bhazelton) [reply pyuvdata/uvdata/uvfits.py:976]: I just made a  new v2.4 milestone and added it there.

File: pyuvdata/uvflag/uvflag.py (hunk 4 above)
Comment (bhazelton) [reply pyuvdata/uvflag/uvflag.py:75]: I just made a  new v2.4 milestone and added it there.

**Pull request merged**